### PR TITLE
Null safety in NewlinesRule

### LINKED ISSUE #554: Newline should be inserted before `?:` in functional call chain
## Describe the bug
```kotlin
fun foo() {
    foo.bar().baz().qux() ?: baz
}
```
should be fixed to
```kotlin
fun foo() {
    foo
        .bar()
        .baz()
        .qux()
        ?: baz
}
```
but `?:` isn't placed on a separate line

## Environment information
* diktat version: 0.1.5

### PR DESCRIPTION
**Pull request by kentr0w**
## Which rule and warnings did you add?
   Fixed bug
   closes #554

### CODE REVIEW

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt [hunk 1]
```diff
@@ -139,6 +130,7 @@ class NewlinesRule(private val configRules: List<RulesConfig>) : Rule("newlines"
         }
     }
 
+    @Suppress("AVOID_NESTED_FUNCTION")
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt:133]: Where is a nested function here?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt [hunk 2]
```diff
@@ -147,7 +139,23 @@ class NewlinesRule(private val configRules: List<RulesConfig>) : Rule("newlines"
 
         // We need to check newline only if prevCodeSibling exists. It can be not the case for unary operators, which are placed
         // at the beginning of the line.
-        if (node.prevCodeSibling()?.isFollowedByNewline() == true) {
+        if (node.firstChildNode != null && node.firstChildNode?.elementType == ELVIS) {
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt:142]:
```suggestion
        if (node.firstChildNode?.elementType == ELVIS) {
```
This check is redundant

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt [hunk 3]
```diff
@@ -147,7 +139,23 @@ class NewlinesRule(private val configRules: List<RulesConfig>) : Rule("newlines"
 
         // We need to check newline only if prevCodeSibling exists. It can be not the case for unary operators, which are placed
         // at the beginning of the line.
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt:141]: This comment should be moved inside the second branch

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt [hunk 4]
```diff
@@ -429,7 +437,9 @@ class NewlinesRule(private val configRules: List<RulesConfig>) : Rule("newlines"
      *
      * @return true - if there is error, false, if there is no error and null if there is two calls in chain
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt:438]:
```suggestion
     * @return true - if there is error, false, if there is no error
```
It doesn't return `null` anymore, does it?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt [hunk 5]
```diff
@@ -147,7 +139,23 @@ class NewlinesRule(private val configRules: List<RulesConfig>) : Rule("newlines"
 
         // We need to check newline only if prevCodeSibling exists. It can be not the case for unary operators, which are placed
         // at the beginning of the line.
-        if (node.prevCodeSibling()?.isFollowedByNewline() == true) {
+        if (node.firstChildNode != null && node.firstChildNode?.elementType == ELVIS) {
+            val binaryExpression = (node.treeParent.psi as KtBinaryExpression)
+            val leftDotCalls = binaryExpression.left?.node
+                ?.takeIf { it.elementType == DOT_QUALIFIED_EXPRESSION }
+                ?.findChildByType(DOT)
+                ?.getCallChain()
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt:147]: These four lines can be extracted to a function and reused for left and right,

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt [hunk 6]
```diff
@@ -509,7 +518,7 @@ class NewlinesRule(private val configRules: List<RulesConfig>) : Rule("newlines"
 
         // fixme: these token sets can be not full, need to add new once as corresponding cases are discovered.
         // error is raised if these operators are prepended by newline
-        private val lineBreakAfterOperators = TokenSet.create(ANDAND, OROR, PLUS, PLUSEQ, MINUS, MINUSEQ, MUL, MULTEQ, DIV, DIVEQ)
+        private val lineBreakAfterOperators = TokenSet.create(ANDAND, OROR, PLUS, PLUSEQ, MINUS, MINUSEQ, MUL, MULTEQ, DIV, DIVEQ, ELVIS)
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt:521]: Whyy? There shouldn't be a line break *after* `?:`, but it still appears in `lineBreakAfterOperators`. Why is this needed?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt (hunk 2 above)
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt:142]: Also, why are you checking it inside `handleOperatorWithLineBreakAfter`? There already is a little bit of special logic for `?:` in `handleOperatorWithLineBreakBefore`, this new logic should be placed there too.

**Review by petertrr — CHANGES_REQUESTED** [review]

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt (hunk 1 above)
Comment (kentr0w) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/NewlinesRule.kt:133]: Forgot to end comment)

**Review by petertrr — APPROVED** [review]

**Pull request merged**